### PR TITLE
build: target Windows 7 when building libevent and fix ipv6 usage

### PR DESCRIPTION
**Pull request by fanquake**
TLDR: This poaches a commit from #18287 and adds one more to adjust the Windows version targeted when building libevent. These changes combined should fully fix ipv6 usage with the RPC server on Windows.

---

Binding the RPC server to a ipv6 address does not currently work on Windows.
We currently try and bind to `127.0.0.1` and `::1` [by default](https://github.com/bitcoin/bitcoin/blob/master/src/httpserver.cpp#L304).

On Windows you'll see lines like this in debug.log:
```bash
2020-06-24T01:49:04Z libevent: getaddrinfo: nodename nor servname provided, or not known
2020-06-24T01:49:04Z Binding RPC on address ::1 port 8332 failed
```

This issue was bought up in, and supposedly fixed by #18287, however the two people that tested it, both said that it didn't fix the problem. I think I now understand why that change alone is incomplete.

Our call into libevent starts with [evhttp_bind_socket_with_handle()](https://github.com/bitcoin/bitcoin/blob/master/src/httpserver.cpp#L325):
```bash
evhttp_bind_socket_with_handle()
	bind_socket()
		make_addrinfo()
			evutil_getaddrinfo()
				if #USE_NATIVE_GETADDRINFO
					#ifndef AI_ADDRCONFIG
						evutil_adjust_hints_for_addrconfig_()
							evutil_check_interfaces()
								evutil_check_ifaddrs()
									evutil_found_ifaddr()
										// miss identifies ipv6 as ipv4?
					#endif
					evutil_getaddrinfo_common_()
```

The problem is falling into ["#ifndef AI_ADDRCONFIG"](https://github.com/libevent/libevent/blob/master/evutil.c#L1580): 

```cpp
#ifndef AI_ADDRCONFIG
	/* Not every system has AI_ADDRCONFIG, so fake it. */
	if (hints.ai_family == PF_UNSPEC &&
	    (hints.ai_flags & EVUTIL_AI_ADDRCONFIG)) {
		evutil_adjust_hints_for_addrconfig_(&hints);
	}
#endif
```

When this occurs, hints end up being adjusted, and it seems that ipv6 addresses end up being mis-identified as ipv4?

However this shouldn't happen, as these `AI_` definitions are available on Windows.
The issue is that in evutil.c, `_WIN32_WINNT` [is set to `0x501`](https://github.com/libevent/libevent/blob/master/evutil.c#L45) (XP).

This obviously predates Vista (`0x0600`), which is when the `AI_ADDRCONFIG` definition (and others) became [available](https://docs.microsoft.com/en-us/windows/win32/api/ws2def/ns-ws2def-addrinfoa).

The change here will override libevents internal D_WIN32_WINNT defines. This should be ok, because it's only making "more" of the Windows API available. It's also aligned with what we do in our own configure, we pass [`D_WIN32_WINNT=0x0601`](https://github.com/bitcoin/bitcoin/blob/master/configure.ac#L610). We also now use linker flags to restrict our binary from running on a Windows version [earlier than Windows 7](https://github.com/bitcoin/bitcoin/blob/master/configure.ac#L621).

The combined fixes can be tested by running:

`bitcoind -rpcbind=::1 rpcallowip='0.0.0.0/0' -debug=http`

and then querying it using:

`bitcoin-cli -rpcconnect=::1 getblockchaininfo`

TODO:
- [x] Open an issue upstream. https://github.com/libevent/libevent/issues/1041

### CODE REVIEW

File: depends/packages/libevent.mk [hunk 1]
```diff
@@ -14,6 +16,7 @@ define $(package)_set_vars
   $(package)_config_opts_release=--disable-debug-mode
   $(package)_config_opts_linux=--with-pic
   $(package)_config_opts_android=--with-pic
+  $(package)_cppflags_mingw32=-D_WIN32_WINNT=0x0601
```
Comment (luke-jr) [inline depends/packages/libevent.mk:19]: I don't see how this can work... evutil.c `#undef _WIN32_WINNT`

Comment (fanquake) [reply depends/packages/libevent.mk:19]: It's only [undef'd](https://github.com/libevent/libevent/blob/45c3fc29fb2240dbaa066c971f3fd59861293305/evutil.c#L43) after the inclusion of [`ws2tcpip.h`](https://github.com/libevent/libevent/blob/45c3fc29fb2240dbaa066c971f3fd59861293305/evutil.c#L33) which is where the `AI_*` definitions are.

In the `ws2tcpip.h` header on my machine I've got:
```c
#define AI_PASSIVE                  0x00000001
#define AI_CANONNAME                0x00000002
#define AI_NUMERICHOST              0x00000004
#if (_WIN32_WINNT >= 0x0600)
#define AI_NUMERICSERV              0x00000008
#define AI_ALL                      0x00000100
#define AI_ADDRCONFIG               0x00000400
#define AI_V4MAPPED                 0x00000800
#define AI_NON_AUTHORITATIVE        0x00004000
#define AI_SECURE                   0x00008000
#define AI_RETURN_PREFERRED_NAMES   0x00010000
#endif
```

**Comment by dahuisong** [comment]
16qwwLpSDngpgkf3cBZbRPW39h9x3dvdMk

File: depends/packages/libevent.mk (hunk 1 above)
Comment (luke-jr) [reply depends/packages/libevent.mk:19]: Oh, weird. That seems like it's probably a bug, but not our problem.

**Comment by luke-jr** [comment]
utACK

**Review by luke-jr — APPROVED** [review]
utACK

**Comment by sipsorcery** [comment]
tACK b440aca96a338f16ca664065821b980c4b0537ae.

Adjusted test commands:
````
bitcoind -regtest -rpcbind=::1 -rpcallowip=0.0.0.0/0 -debug=http
bitcoin-cli -regtest -rpcconnect=::1 getblockchaininfo
````

Relevant parts of bitcoind console messages:
````
....
2020-06-25T10:21:59Z Command-line arg: debug="http"
2020-06-25T10:21:59Z Command-line arg: regtest=""
2020-06-25T10:21:59Z Command-line arg: rpcallowip="0.0.0.0/0"
2020-06-25T10:21:59Z Command-line arg: rpcbind=****
2020-06-25T10:21:59Z Using at most 125 automatic connections (2048 file descriptors available)
2020-06-25T10:21:59Z Using 16 MiB out of 32/2 requested for signature cache, able to store 524288 elements
2020-06-25T10:21:59Z Using 16 MiB out of 32/2 requested for script execution cache, able to store 524288 elements
2020-06-25T10:21:59Z Script verification uses 7 additional threads
2020-06-25T10:21:59Z Allowing HTTP connections from: 127.0.0.0/8 ::1/128 0.0.0.0/0
2020-06-25T10:21:59Z scheduler thread start
2020-06-25T10:21:59Z Binding RPC on address ::1 port 18443
2020-06-25T10:21:59Z Initialized HTTP server
....
2020-06-25T10:23:07Z Received a POST request for / from [::1]:52391
````

**Review by sipsorcery — APPROVED** [review]

**Comment by fanquake** [comment]
Just a clarification to "how" this is working, and follow up to the comments above. By default, the mingw32 compiler, (on my machine, `x86_64-w64-mingw32-g++ (GCC) 9.3.0`) will set `_WIN32_WINNT` to 0x0502 ([Windows Server 2003](https://docs.microsoft.com/en-us/cpp/porting/modifying-winver-and-win32-winnt?view=vs-2019)).

So the fix here isn't just that we are overriding libevents internal `_WIN32_WINNT` defines, it's because we are actually settings a value for `_WIN32_WINNT`, which overrides the default set by the compiler, and, is high enough that the definitions are available.

The seconds point is header include order. Because the `ws2tcpip.h` header, which [contains the `AI_*` definitions](https://github.com/mirror/mingw-w64/blob/master/mingw-w64-headers/include/ws2tcpip.h#L222):
```c
#define AI_PASSIVE                  0x00000001
#define AI_CANONNAME                0x00000002
#define AI_NUMERICHOST              0x00000004
#if (_WIN32_WINNT >= 0x0600)
#define AI_NUMERICSERV              0x00000008
#define AI_ALL                      0x00000100
#define AI_ADDRCONFIG               0x00000400
#define AI_V4MAPPED                 0x00000800
#define AI_NON_AUTHORITATIVE        0x00004000
#define AI_SECURE                   0x00008000
#define AI_RETURN_PREFERRED_NAMES   0x00010000
#endif
```
is included _before_ libevent #undefs `_WIN32_WINNT` and sets it to `0x0501`, `AI_ADDRCONFIG` et al will be defined, because at that point, _WIN32_WINNT is set to the value we have passed (`0x0601`).

File: depends/packages/libevent.mk (hunk 1 above)
Comment (laanwj) [reply depends/packages/libevent.mk:19]: You might want to add a comment here in the makefile about this is defined, and why it works.

Comment (fanquake) [reply depends/packages/libevent.mk:22]: @laanwj I've added a comment here. These kinds of details probably deserve a separate document in depends (we could also add some of the macOS stuff there as well). I can follow up by writing up some more extensive docs.

**Comment by laanwj** [comment]
ACK eb6b73540d1ee7ff5a6874dd0e35f9b30b68e3b8

**Comment by theuni** [comment]
Post-merge ACK eb6b73540d1ee7ff5a6874dd0e35f9b30b68e3b8 .
